agent/eni/watcher.go

```diff
@@ -0,0 +1,255 @@
+func (udevWatcher *UdevWatcher) Init() error {
+	links, err := udevWatcher.netlinkClient.LinkList()
+	if err != nil {
+		return errors.Wrapf(err, "init udevWatcher: error retrieving network interfaces")
+	}
+
+	udevWatcher.updateLock.Lock()
+	for _, link := range links {
+		deviceName := link.Attrs().Name
+		macAddress := link.Attrs().HardwareAddr.String()
+		if macAddress != "" {
+			udevWatcher.addDeviceWithMACAddressUnsafe(deviceName, macAddress)
+		}
+	}
+	udevWatcher.updateLock.Unlock()
+
+	// UDev Event Handler
+	go udevWatcher.eventHandler(udevWatcher.ctx)
+	return nil
+}
+
+// Start periodically updates the state of ENIs connected to the system
+func (udevWatcher *UdevWatcher) Start() {
+	udevWatcher.performPeriodicReconciliation(udevWatcher.ctx, defaultReconciliationInterval)
+}
+
+// Stop is used to invoke the cancellation routine
+func (udevWatcher *UdevWatcher) Stop() {
+	udevWatcher.cancel()
+}
+
+// performPeriodicReconciliation is used to periodically invoke the
+// reconciliation process based on a ticker
+func (udevWatcher *UdevWatcher) performPeriodicReconciliation(ctx context.Context, updateInterval time.Duration) {
+	udevWatcher.updateIntervalTicker = time.NewTicker(updateInterval)
+	for {
+		select {
+		case <-udevWatcher.updateIntervalTicker.C:
+			udevWatcher.reconcileOnce()
+		case <-ctx.Done():
+			udevWatcher.updateIntervalTicker.Stop()
+			return
+		}
+	}
+}
+
+// reconcileOnce is used to reconcile the state of ENIs attached to the instance
+func (udevWatcher *UdevWatcher) reconcileOnce() {
+	links, err := udevWatcher.netlinkClient.LinkList()
+	if err != nil {
+		log.Warnf("Udev watcher reconciliation: error retrieving network interfaces: %v", err)
+		return
+	}
+
+	// Return on empty list
+	if len(links) == 0 {
+		log.Info("Udev watcher reconciliation: no network interfaces discovered for reconciliation")
+		return
+	}
+
+	currentState := udevWatcher.buildState(links)
+
+	// NOTE: For correct semantics, this entire function needs to be locked.
+	// As we postulate the netlinkClient.LinkList() call to be expensive, we allow
+	// the race here. The state would be corrected during the next reconciliation loop.
+	udevWatcher.updateLock.Lock()
+	defer udevWatcher.updateLock.Unlock()
+
+	// Remove non-existent interfaces first
+	for managedMACAddress, managedDeviceName := range udevWatcher.enis {
+		if currentDeviceName, ok := currentState[managedMACAddress]; !ok || managedDeviceName != currentDeviceName {
+			udevWatcher.removeDeviceWithMACAddressUnsafe(managedMACAddress)
+		}
+	}
+
+	// Add new interfaces next
+	for mac, dev := range currentState {
+		if _, ok := udevWatcher.enis[mac]; !ok && mac != "" {
+			udevWatcher.addDeviceWithMACAddressUnsafe(dev, mac)
+		}
+	}
+}
+
+// getAllENIs is used to retrieve the state observed by the Watcher
+func (udevWatcher *UdevWatcher) getAllENIs() map[string]string {
+	udevWatcher.updateLock.RLock()
+	defer udevWatcher.updateLock.RUnlock()
+	return udevWatcher.enis
```

Comment:
I think there maybe some problem here, since `enis` is just a reference to the map, the map can still be changed  while reading in line 171. Please correct me if I misunderstood.